fabtests/hpcs/src/core/core.c

```diff
@@ -0,0 +1,1476 @@
+{
+	uint64_t my_key;
+	int i, ret;
+
+	if (args->test_api.rx_create_mr == NULL || !config->rx_use_mr)
+		return 0;
+
+	if (args->window_size < job->ranks) {
+		hpcs_error("for one-sided communication, window must be >= number of ranks\n");
+		return (-FI_EINVAL);
+	}
+
+	/* Key can be any arbitrary number. */
+	*rx_mr = args->test_api.rx_create_mr(args->test_arguments, domain_state->domain, 42+job->rank,
+			rx_buf, args->window_size*args->buffer_size, config->mr_rx_flags, 0);
+	if (*rx_mr == NULL) {
+		hpcs_error("failed to create target memory region\n");
+		return -1;
+	}
+
+	my_key = fi_mr_key(*rx_mr);
+	job->address_exchange(&my_key, keys, sizeof(uint64_t), job->ranks);
+
+	if (verbose) {
+		hpcs_verbose("mr key exchange complete: rank %ld my_key %ld len %ld keys: ",
+				job->rank, my_key, args->window_size*args->buffer_size);
+		for (i=0; i < job->ranks; i++)
+			printf("%ld ", keys[i]);
+		printf("\n");
+	}
+
+	if (config->rx_use_cntr) {
+		if (!domain_state->rx_cntr) {
+			hpcs_error("no rx counter to bind mr to\n");
+			return -FI_EINVAL;
+		}
+
+		ret = fi_mr_bind(*rx_mr, &domain_state->rx_cntr->fid, config->mr_rx_flags);
+		if (ret) {
+			hpcs_error("fi_mr_bind (rx_cntr) failed: %d\n", ret);
+
+			/*
+			 * Binding an MR with FI_REMOTE_READ isn't defined by the OFI spec,
+ 			 * so we don't consider this a failure.
+			 */
+			if (config->mr_rx_flags & FI_REMOTE_READ) {
+					hpcs_error("FI_REMOTE_READ memory region bind flag unsupported by this provider, skipping test.\n");
+				return -FI_EOPNOTSUPP;
+			}
+
+			return -1;
+		}
+	}
+
+	ret = fi_mr_enable(*rx_mr);
+	if (ret)
+		hpcs_error("fi_mr_enable failed: %d\n", ret);
+
+	job->barrier();
+
+	return 0;
+}
+
+/* Core loop progress information and context state. */
+struct core_state {
+	/* iteration counter (sometimes useful for debug messages) */
+	size_t			iteration;
+
+	/* allocated and pre-initialized memory resources */
+	uint8_t			*rx_buf;
+	uint8_t			*tx_buf;
+	struct op_context	*tx_context;
+	struct op_context	*rx_context;
+	uint64_t		*keys;
+	struct fid_mr		*rx_mr;
+
+	/* initiated and completed operation counters, not reset per iteration */
+	size_t			recvs_posted;
+	size_t			sends_posted;
+	size_t			recvs_done;
+	size_t			sends_done;
+
+	/* sends/recvs completed at beginning of current iteration */
+	size_t			sends_done_prev;
+	size_t			recvs_done_prev;
+
+	/* window slots */
+	size_t			tx_window;
+	size_t			rx_window;
+
+	/* pattern iterator state */
+	int			cur_sender;
+	int			cur_receiver;
+	int			cur_sender_rx_threshold;
+	int			cur_receiver_tx_threshold;
+
+	/* current iteration is complete when all three are true */
+	bool			all_recvs_done;
+	bool			all_sends_done;
+	bool			all_completions_done;
+
+	/* options */
+	uint64_t		tx_flags;
+	uint64_t		rx_flags;
+};
+
+
+#define DATA_BUF(base, counter) \
+		((base) + ((counter % arguments->window_size) * arguments->buffer_size))
+
+#define CONTEXT(base, counter) (&base[counter % arguments->window_size])
```

Comment:
replace defines with static functions for type checking

Reply:
They're #defines because they pull in arguments from the larger context.

However, I suppose we don't call them often enough that the brevity is worth having macros with non-intuitive semantics.